File: euler_092.py
from time import time

start = time()

squares = {n:n * n for n in range(600)}

sd = {n:0 for n in range(10 ** 7)}
for n in range(10):
    sd[n] = squares[n]

n = 10

while n < 10 ** 7:
    sd[n] = squares[n % 10] + sd[n // 10]
    n += 1

eighty_niners = 0

for m in range(2, 10 ** 7):
    next_num = sd[m]
    while not next_num in [1, 89]:
        next_num = sd[next_num]
    if next_num == 89:
        eighty_niners += 1
        continue

print("Number of chains ending in 89: ", eighty_niners)

stop = time()
print("Time: " + str(stop - start))

File: euler_092.js
const start = Date.now();

let squares = [];
for (let i = 0; i < 10; i++) {
    squares[i] = i * i;
}

const nextNumber = (cur) => {
    if (cur < 10) {
        return squares[cur];
    }
    return squares[cur % 10] + nextNumber(Math.floor(cur / 10));
}

let checked = new Array(10**7);
checked.fill(0);

for (let i = 2; i < 10**7; i++) {
    if (checked[i] === 0) {
        let chain = [i];
        let cur = nextNumber(i);
        while (cur !== 1 && cur !== 89) {
            chain.push(cur);
            cur = nextNumber(cur);
        }
        for (const c of chain) {
            checked[c] = cur;
        }
    }
}

const eightyNines = checked.filter(x => x === 89).length;

const end = Date.now();

console.log("Execute time: ", end - start, " The number of starting numbers is: ", eightyNines);